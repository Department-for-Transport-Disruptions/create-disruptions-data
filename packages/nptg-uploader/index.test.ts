import { promises as fs } from "fs";
import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
import * as nptgUploader from "./index";

let nptgString = "";

describe("nptg-uploader", () => {
    const hoisted = vi.hoisted(() => ({
        dropTableMock: vi.fn().mockImplementation(() => ({
            ifExists: () => ({
                execute: vi.fn(),
            }),
        })),
        valuesMock: vi.fn().mockImplementation(() => ({
            execute: vi.fn().mockResolvedValue(null),
        })),
        insertMock: vi.fn().mockImplementation(() => ({
            values: hoisted.valuesMock,
        })),
    }));

    beforeAll(async () => {
        nptgString = await fs.readFile(`${__dirname}/test-data/nptg.xml`, "utf-8");
    });

    beforeEach(() => {
        vi.mock("@create-disruptions-data/shared-ts/utils/db", () => ({
            getDbClient: () => ({
                schema: {
                    dropTable: hoisted.dropTableMock,
                },
                sql: vi.fn(),
                getExecutor: () => ({
                    executeQuery: vi.fn(),
                    transformQuery: vi.fn(),
                    compileQuery: vi.fn(),
                }),
                executorProvider: {
                    getExecutor: vi.fn(),
                },
                insertInto: hoisted.insertMock,
            }),
        }));
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    it("uploads correct admin areas and localities to the database", async () => {
        await nptgUploader.parseNptgAndUpload(nptgString);

        expect(hoisted.insertMock).toBeCalledTimes(2);

        expect(hoisted.insertMock.mock.calls[0]).toStrictEqual(["nptgAdminAreasNew"]);
        expect(hoisted.valuesMock.mock.calls[0]).toMatchSnapshot();

        expect(hoisted.insertMock.mock.calls[1]).toStrictEqual(["localitiesNew"]);
        expect(hoisted.valuesMock.mock.calls[1]).toMatchSnapshot();
    });
});
