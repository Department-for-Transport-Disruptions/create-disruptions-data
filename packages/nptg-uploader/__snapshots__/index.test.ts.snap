// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nptg-uploader > uploads correct admin areas and localities to the database 1`] = `
[
  [
    {
      "administrativeAreaCode": "082",
      "atcoAreaCode": "490",
      "name": "Greater London",
      "shortName": "London",
    },
    {
      "administrativeAreaCode": "015",
      "atcoAreaCode": "076",
      "name": "Darlington",
      "shortName": "Darlington",
    },
    {
      "administrativeAreaCode": "022",
      "atcoAreaCode": "075",
      "name": "Hartlepool",
      "shortName": "Hartlepool",
    },
    {
      "administrativeAreaCode": "031",
      "atcoAreaCode": "079",
      "name": "Middlesbrough",
      "shortName": "Middlesbrough",
    },
    {
      "administrativeAreaCode": "047",
      "atcoAreaCode": "078",
      "name": "Redcar & Cleveland",
      "shortName": "Cleveland",
    },
    {
      "administrativeAreaCode": "054",
      "atcoAreaCode": "077",
      "name": "Stockton-on-Tees",
      "shortName": "Stockton",
    },
    {
      "administrativeAreaCode": "078",
      "atcoAreaCode": "130",
      "name": "Durham",
      "shortName": "Durham",
    },
    {
      "administrativeAreaCode": "094",
      "atcoAreaCode": "310",
      "name": "Northumberland",
      "shortName": "Northumberland",
    },
    {
      "administrativeAreaCode": "103",
      "atcoAreaCode": "410",
      "name": "Tyne & Wear",
      "shortName": "Tyne & Wear",
    },
  ],
]
`;

exports[`nptg-uploader > uploads correct admin areas and localities to the database 2`] = `
[
  [
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T11:18:07",
      "easting": 365491,
      "gridType": null,
      "localityName": "Amesbury",
      "localityNameLang": "EN",
      "modification": "revise",
      "modificationDateTime": "2021-08-17T11:15:30.703",
      "northing": 158651,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034964",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": "Bath & North East Somerset",
      "qualifierNameLang": "en",
      "revisionNumber": 1,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T11:18:23",
      "easting": 370798,
      "gridType": null,
      "localityName": "Ashgrove",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T11:18:23",
      "northing": 157864,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034965",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": "Peasedown St John",
      "qualifierNameLang": "en",
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:03",
      "easting": 376748,
      "gridType": null,
      "localityName": "Bailbrook",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:03",
      "northing": 167119,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034968",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:09",
      "easting": 364747,
      "gridType": null,
      "localityName": "Barrow Vale",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:09",
      "northing": 160235,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034970",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:11",
      "easting": 375615,
      "gridType": null,
      "localityName": "Bathwick",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:11",
      "northing": 165218,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034971",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T11:20:33",
      "easting": 375140,
      "gridType": null,
      "localityName": "Beacon Hill",
      "localityNameLang": "EN",
      "modification": "revise",
      "modificationDateTime": "2018-03-19T15:19:58.680",
      "northing": 166178,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034972",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": "Bath",
      "qualifierNameLang": "en",
      "revisionNumber": 1,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:14",
      "easting": 374896,
      "gridType": null,
      "localityName": "Beechen Cliff",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:14",
      "northing": 164108,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034973",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:15",
      "easting": 361290,
      "gridType": null,
      "localityName": "Belluton",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:15",
      "northing": 164420,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034974",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
    {
      "administrativeAreaCode": "001",
      "creationDateTime": "2005-10-05T10:45:22",
      "easting": 358644,
      "gridType": null,
      "localityName": "Bishop Sutton",
      "localityNameLang": "EN",
      "modification": "new",
      "modificationDateTime": "2005-10-05T10:45:22",
      "northing": 159607,
      "nptgDistrictCode": "310",
      "nptgLocalityCode": "E0034975",
      "parentLocalityName": null,
      "parentLocalityNameLang": null,
      "qualifierDistrictRef": null,
      "qualifierLocalityRef": null,
      "qualifierName": null,
      "qualifierNameLang": null,
      "revisionNumber": 0,
      "shortName": null,
      "shortNameLang": null,
      "sourceLocalityType": "Lo",
    },
  ],
]
`;
